python/ray/autoscaler/v2/tests/test_e2e.py

```diff
@@ -200,6 +207,111 @@ def verify():
+
+    obj = ray.put("hello")
+
+    def verify():
+        state = get_cluster_status()
+        for node in state.healthy_nodes:
+            assert node.node_status == "RUNNING"
+            assert node.used_resources()["object_store_memory"] > 0
+            return True
+
+    wait_for_condition(verify)
+
+    del obj
+
+    import time
+
+    time.sleep(1)
+
+    def verify():
+        state = get_cluster_status()
+        for node in state.healthy_nodes:
+            assert node.node_status == "IDLE"
+            assert node.used_resources()["object_store_memory"] == 0
+            assert node.resource_usage.idle_time_ms >= 1000
+            return True
+
+    wait_for_condition(verify)
+
+
+def test_serve_num_replica_idle_node():
+    # Test that nodes become idle after serve scaling down.
+    cluster = AutoscalingCluster(
+        head_resources={"CPU": 0},
+        worker_node_types={
+            "type-1": {
+                "resources": {"CPU": 4},
+                "node_config": {},
+                "min_workers": 0,
+                "max_workers": 30,
+            },
+        },
+        idle_timeout_minutes=1,
+    )
+
+    from ray import serve
+
+    @serve.deployment(ray_actor_options={"num_cpus": 2})
+    class Deployment:
+        def __call__(self):
+            return "hello"
+
+    try:
+        cluster.start(override_env={"RAY_SERVE_PROXY_MIN_DRAINING_PERIOD_S": "2"})
+        # 5 workers nodes should be busy and have 2(replicas) * 2(cpus per replicas)
+        # = 4 cpus used
+        serve.run(Deployment.options(num_replicas=10).bind())
+
+        expected_num_workers = 5
+
+        def verify():
+            cluster_state = get_cluster_status()
+
+            # Verify that nodes are busy.
+            assert len((cluster_state.healthy_nodes)) == expected_num_workers + 1
+            for node in cluster_state.healthy_nodes:
+                assert node.node_status == "RUNNING"
+                if not is_head_node_from_resource_usage(node.total_resources()):
+                    available = node.available_resources()
+                    assert available["CPU"] == 0
+            return True
+
+        wait_for_condition(verify)
+
+        # Downscale to 1 replicas, 2 workers nodes should be idle.
+        serve.run(Deployment.options(num_replicas=2).bind())
+
+        def verify():
+            cluster_state = get_cluster_status()
+
+            assert len((cluster_state.healthy_nodes)) == expected_num_workers + 1
+            idle_nodes = []
+
+            for node in cluster_state.healthy_nodes:
+                if not is_head_node_from_resource_usage(node.total_resources()):
+                    available = node.available_resources()
+                    if node.node_status == "IDLE":
+                        assert available["CPU"] == 4
+                        idle_nodes.append(node)
+            from rich import print
+
+            print(cluster_state.healthy_nodes)
+            assert len(idle_nodes) == 3
```

Comment:
3 depends on that the remaining two replicas are on two nodes, which is not guaranteed. It's better to downscale to 1 replica and verify that 4 nodes are idle.

Reply:
sg